### PR TITLE
fix: Fix gas fee polling not being stopped for popup UI instances in chrome

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

Currently extension popups in chrome do not properly clean up react effects. Previously, gas fee polling started from UI components were also tracked by MetaMaskController and cleaned up when the controller lost connection with the popup (or fullscreen, or notification) UI instances. The new gas fee polling hook failed to account for this and did not properly track its polling tokens with background which introduced the possibility of gas fee polling not stopping correctly in certain instances. This PR fixes this by updating the `usePolling` hook to add and remove poll tokens from app state and also updates MetaMaskController to stop any active gas fee polling based on the polling tokens in app state.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23594?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2014

## **Manual testing steps**

In chrome

1. Open a popup instance of extension (click the extension icon from the toolbar)
2. Open background console network tab
3. Get to the send transaction confirmation screen
4. Verify gas fee requests are happening
5. Close the popup instance by unfocusing it
6. Verify gas fee requests have stopped

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/send/send-content/__snapshots__/send-content.component.test.js.snap [hunk 1]
```diff
@@ -1,5 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`SendContent Component SendHexDataRow should not render the SendHexDataRow if props.showHexData is false 1`] = `<div />`;
```
Comment (Gudahtt) [inline ui/pages/confirmations/send/send-content/__snapshots__/send-content.component.test.js.snap:3]: Is this change intentional?

File: ui/hooks/usePolling.ts [hunk 2]
```diff
@@ -20,6 +24,7 @@ const usePolling = (usePollingOptions: UsePollingOptions) => {
 
     const cleanup = () => {
       if (pollTokenRef.current) {
+        removePollingTokenFromAppState(pollTokenRef.current);
```
Comment (Gudahtt) [inline ui/hooks/usePolling.ts:27]: Do we ever want to stop tracking a token without stopping polling? Or the other way around - stop polling without removing a token from the app state? This seems like it should be one single function, rather than two.

File: app/scripts/metamask-controller.js [hunk 3]
```diff
@@ -5687,6 +5688,7 @@ export default class MetamaskController extends EventEmitter {
       this.appStateController.store.getState()[appStatePollingTokenType];
     pollingTokensToDisconnect.forEach((pollingToken) => {
       this.gasFeeController.disconnectPoller(pollingToken);
+      this.gasFeeController.stopPollingByPollingToken(pollingToken);
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:5691]: It looks like this is the new polling API, whereas `disconnectPoller` is the old API. Is that correct? Are we using both of these simultaneously?

File: ui/hooks/usePolling.ts (hunk 2 above)
Comment (jiexi) [reply ui/hooks/usePolling.ts:27]: Ahh i see. Yeah they should be coupled in another function

File: app/scripts/metamask-controller.js (hunk 3 above)
Comment (jiexi) [reply app/scripts/metamask-controller.js:5691]: there are a few references to the old API still

File: ui/pages/confirmations/send/send-content/__snapshots__/send-content.component.test.js.snap (hunk 1 above)
Comment (jiexi) [reply ui/pages/confirmations/send/send-content/__snapshots__/send-content.component.test.js.snap:3]: no, this seems to be that bug we are seeing where local and CI expect different results for this component's snapshot

File: ui/hooks/usePolling.ts (hunk 2 above)
Comment (jiexi) [reply ui/hooks/usePolling.ts]: done here: https://github.com/MetaMask/metamask-extension/pull/23594/commits/3b4fb8f0355bc5f50430d25d6b1cfe5cbbfbca56

Comment (jiexi) [reply ui/hooks/usePolling.ts]: apologies. done here instead https://github.com/MetaMask/metamask-extension/pull/23594/commits/029f53e1be6275aa9977f9d054854381dc3e981e

made more sense to couple in the actions

**Review by shanejonas — APPROVED** [review]

File: app/scripts/metamask-controller.js (hunk 3 above)
Comment (adonesky1) [reply app/scripts/metamask-controller.js:5678]: @jiexi can you remind me where these references are? These use the same polling tokens?

Comment (jiexi) [reply app/scripts/metamask-controller.js:5678]: https://github.com/MetaMask/metamask-extension/blob/0a3ef253bbec5c3245078d9bbbf166fbe8b32982/ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js#L43

https://github.com/MetaMask/metamask-extension/blob/0a3ef253bbec5c3245078d9bbbf166fbe8b32982/ui/ducks/send/send.js#L53

Comment (jiexi) [reply app/scripts/metamask-controller.js:5678]: the poll token from the legacy polling and the networkClientId based polling are not the same

Comment (jiexi) [reply app/scripts/metamask-controller.js:5678]: correct. calling `disconnectPoller` and `stopPollingByPollingToken` with polling tokens it cannot find is a noop without errors thrown

**Review by adonesky1 — APPROVED** [review]
LGTM!
